onnxruntime/core/providers/stvm/stvm_execution_provider.cc

```diff
@@ -0,0 +1,538 @@
+
+  ONNX_NAMESPACE::ModelProto model_proto = model.ToProto();
+  model_proto.set_ir_version(ONNX_NAMESPACE::Version::IR_VERSION);
+  auto status = graph_build.Resolve();
+
+  std::unordered_set<std::string> required_initializers;
+  const std::vector<NodeIndex>& sorted_nodes = graph_viewer.GetNodesInTopologicalOrder();
+  std::unique_ptr<IndexedSubGraph> sub_graph = std::make_unique<IndexedSubGraph>();
+  for (auto& node_idx : sorted_nodes) {
+    graph_viewer.GetNode(node_idx)->ForEachDef([&required_initializers, &init_tensors](const NodeArg& node_arg, bool is_input) {
+              if(is_input && init_tensors.count(node_arg.Name())) {
+                  required_initializers.insert(node_arg.Name());
+              } }, true);
+  }
+
+  auto meta_def = std::make_unique<::onnxruntime::IndexedSubGraph::MetaDef>();
+  meta_def->name = "TVMStandalone";
+  meta_def->domain = "StandaloneTest";
+  std::vector<std::string> inputs;
+  std::vector<std::string> outputs;
+
+  for (auto& nodeArgPtr : graph_viewer.GetInputs()) {
+    inputs.push_back(nodeArgPtr->Name());
+  }
+
+  for (auto& name : required_initializers) {
+    inputs.push_back(name);
+  }
+
+  for (auto& nodeArgPtr : graph_viewer.GetOutputs()) {
+    outputs.push_back(nodeArgPtr->Name());
+  }
+  meta_def->inputs = inputs;
+  meta_def->outputs = outputs;
+  meta_def->since_version = 1;
+  meta_def->status = ONNX_NAMESPACE::EXPERIMENTAL;
+  sub_graph->SetMetaDef(std::move(meta_def));
+  sub_graph->nodes = sorted_nodes;
+  result.push_back(
+      std::make_unique<ComputeCapability>(std::move(sub_graph)));
+  return result;
+}
+
+common::Status StvmExecutionProvider::Compile(const std::vector<Node*>& nodes,
+                                              std::vector<NodeComputeInfo>& node_compute_funcs) {
+  for (auto* fused_node : nodes) {
+    auto func_body = fused_node->GetFunctionBody();
+    if (!func_body)
+      return common::Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT, "Function body is empty");
+    const std::string func_name = fused_node->Name();
+    const Graph& node_graph = func_body->Body();
+    Model model(node_graph.Name(), true, ModelMetaData(), PathString(),
+                             IOnnxRuntimeOpSchemaRegistryList(), node_graph.DomainToVersionMap(),
+                             std::vector<ONNX_NAMESPACE::FunctionProto>(), *GetLogger());
+    ONNX_NAMESPACE::ModelProto model_proto = model.ToProto();
+
+    model_proto.set_ir_version(ONNX_NAMESPACE::Version::IR_VERSION);
+    *(model_proto.mutable_graph()) = node_graph.ToGraphProto();
+    auto opset = model_proto.add_opset_import();
+    opset->set_domain(kOnnxDomain);
+    opset->set_version(node_graph.DomainToVersionMap().at(kOnnxDomain));
+
+    std::string string_buf;
+    model_proto.SerializeToString(&string_buf);
+    buffers_[func_name] = string_buf;
+    opsets_[func_name] = int(opset->version());
+    model_paths_[func_name] = fused_node->ModelPath().ToPathString();;
```

Comment:
A bit unclear why these subgraph nodes would have a model-path. Is the model-path optional or required? May be worth checking if the model-path does exist.

Reply:
First point.
The onnxruntime project has a native C ++ part and a python part. To carry out communication between these two parts, google protobuf is used. Data transfer between these parts is done by serializing the .onnx model (serialization to a string in our case). But google protobuf has a model size limit which is 2GB. In view of this, for large models (the size of which exceeds 2 GB) storage is a little different, there is a separate .onnx file that contains the model graph, and several .weight, .data files that contain weights and external initializers for the graph, usually these files are in the same directory. Therefore, to work with such models, it is necessary to load external data for the loaded graph from the .onnx file.
Second point.
Working with the model inference in ONNX Runtime is done by creating an InferenceSession object. This object can be created in two ways:  based on the serialized model when initializing an object by byte array and based on the path to the model.

The problem is that when using a byte array, the model is preloaded using the ONNX API, which by default additionally loads external data that may be required for the onnx model. But if you use only the path to the model, then inside the onnx runtime you need to load external data yourself. The problem is that to use this ONNX API, you need to know the path to external data, but at the time of calling the required function, we have the onnx model serialized into a string and we do not know where the external data for it is. Therefore, we save the path to the model in order to later load external data, if it exists.

Checking that the path to the model is correct (if passed) is carried out inside the ORT, because the model is loaded there for subsequent work with it.